block-storage/src/main/scala/coop/rchain/blockstorage/util/io/LmdbDbi.scala

```diff
@@ -0,0 +1,65 @@
+package coop.rchain.blockstorage.util.io
+
+import java.io.{IOException, PrintWriter, StringWriter}
+
+import cats.implicits._
+import cats.effect.{ExitCase, Sync}
+import coop.rchain.blockstorage.util.io.IOError.RaiseIOError
+import coop.rchain.shared.Log
+import coop.rchain.shared.Resources.withResource
+import org.lmdbjava._
+
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+final case class LmdbDbi[F[_]: Sync: Log: RaiseIOError, T](
+    env: Env[T],
+    dbi: Dbi[T]
+) {
+  private[this] def withTxn[R](txnThunk: => Txn[T])(f: Txn[T] => R): F[R] =
+    Sync[F].bracketCase(Sync[F].delay(txnThunk)) { txn =>
+      Sync[F].delay {
+        val r = f(txn)
+        txn.commit()
+        r
+      }
+    } {
+      case (txn, ExitCase.Error(NonFatal(ex))) =>
+        val stringWriter = new StringWriter()
+        ex.printStackTrace(new PrintWriter(stringWriter))
```

Comment:
what's the gain of allocating a `StringWriter`?

Reply:
Printing the stack trace in logs instead of stderr

Reply:
isn't this enough?
`def error(msg: => String, cause: Throwable)(implicit ev: LogSource): F[Unit]`